### PR TITLE
Add query builder header in embedded dialog

### LINKED ISSUE #3830: Add the ability to switch from detailed to basic view in query dialogs
image

At this time, there is no way to switch to the basic view, which would be very useful for these dialogs in particular as edits are not being made.

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #3830

### CODE REVIEW

File: specifyweb/frontend/js_src/lib/components/QueryBuilder/Header.tsx [hunk 1]
```diff
@@ -61,6 +61,7 @@ export function QueryHeader({
   readonly unsetUnloadProtect: () => void;
   readonly onTriedToSave: () => void;
   readonly onSaved: () => void;
+  readonly toggleMapping: () => void;
```
Comment (maxxxxxdlp) [inline specifyweb/frontend/js_src/lib/components/QueryBuilder/Header.tsx:64]: `onToggleMapping`?
to be consistent with how we call functions that mutate state

File: specifyweb/frontend/js_src/lib/components/QueryBuilder/reducer.ts [hunk 2]
```diff
@@ -52,6 +53,7 @@ export const getInitialState = ({
 }): MainState => ({
   type: 'MainState',
   fields: parseQueryFields(query.fields ?? []),
+  showMappingView: getCache('queryBuilder', 'showMappingView') ?? true,
```
Comment (maxxxxxdlp) [inline specifyweb/frontend/js_src/lib/components/QueryBuilder/reducer.ts:56]: would it be possible to call `useCachedState` in the components that need this cache instead and remove `showMappingView` from the MainState?

putting getCache/setState in a state is a bad pattern I used when I was designing first react components in Specify, but don't do that anymore. reasons: central state is more verbose (requires more boilerplate code), leads to worse performance (any change to the state triggers re-render of all components that use the state in any way) and is bug prone (useCachedState listens for cache being updated by some other component somewhere else. getCache/setCache does not listen, thus the state could get out of date)

**Review by maxxxxxdlp — CHANGES_REQUESTED** [review]

**Comment by maxxxxxdlp** [comment]
@specify/ux-testing when testing, please be extra careful to also test the save/save as buttons
test in different places where embedded query appears: Search Dialog, Meta->Edit History, (for a pick list record) Meta->Pick List Usages

File: specifyweb/frontend/js_src/lib/components/QueryBuilder/Header.tsx (hunk 1 above)
Comment (CarolineDenis) [reply specifyweb/frontend/js_src/lib/components/QueryBuilder/Header.tsx:64]: just took the name that was in production but can change to onToggleMapping if it's better

File: specifyweb/frontend/js_src/lib/components/QueryBuilder/reducer.ts (hunk 2 above)
Comment (CarolineDenis) [reply specifyweb/frontend/js_src/lib/components/QueryBuilder/reducer.ts]: I pushed a code using cachedState but not sure that is what you meant....

File: specifyweb/frontend/js_src/lib/components/QueryBuilder/Header.tsx [hunk 3]
```diff
@@ -79,6 +80,11 @@ export function QueryHeader({
 
   const [isBasic, setIsBasic] = useQueryViewPref(query.id);
 
+  const [showMappingView = true, _] = useCachedState(
```
Comment (maxxxxxdlp) [inline specifyweb/frontend/js_src/lib/components/QueryBuilder/Header.tsx:83]: you don't need to pass down `onToggleMapping` at all

the 2nd argument of `useCachedState` in here is the callback to change the state - thus you can change the state from here, without the need for any props from the parent

remember that changes to `useCachedState` are synchronized between all the places where `useCachedState` is used

**Review by maxxxxxdlp — CHANGES_REQUESTED** [review]

**Comment by CarolineDenis** [comment]
@maxxxxxdlp all the query builder are synchronized atm, if we click Hide Mapping in one query it will be the case for all the queries. Is that what we want?

**Comment by melton-jason** [comment]
> @maxxxxxdlp all the query builder are synchronized atm, if we click Hide Mapping in one query it will be the case for all the queries. Is that what we want?

I believe that would also be a question to ask @specify/ux-testing :wink:

**Review by maxxxxxdlp — APPROVED** [review]
remove commented out code

**Comment by maxxxxxdlp** [comment]
> @maxxxxxdlp all the query builder are synchronized atm, if we click Hide Mapping in one query it will be the case for all the queries. Is that what we want?

Fine with me either way.
If you want, you can store in separate local storage keys depending on whether query builder is embedded or not - as most users would likely want to see the mapping explorer in the full query builder, but might want to hide it when in the dialog if on a smaller screen

**Review by bronwyncombs — CHANGES_REQUESTED** [review]
When saving a query made in an embedded dialog (not via stats) it will prompt you to leave the current form and open the saved query in the main query builder.

It should just save the query and not show this dialog. Same applies to Save As button. 

https://github.com/specify/specify7/assets/135047322/ec58a46b-3c7b-4ead-927d-e0fcfe895f6b

---
**Comments:**

Embedded dialog header on stats page looks good! Switching to basic view reduces the width of the dialog, but I think this is fine. 

Embedded and full query builders in sync so detailed/basic view is carried over

**Comment by CarolineDenis** [comment]
Decision: 
Keep "Basic View" and "Hide Field Mapper" bins in embedded header dialog. Remove "Save Query" btn from it

**Review by bronwyncombs — APPROVED** [review]
Looks good! Save/Save As issue resolved.

**Comment by maxxxxxdlp** [comment]
For future reference, we could bring back the save button, or fix https://github.com/specify/specify7/issues/3000

**Pull request merged**